src/program/reflection.rs

```diff
@@ -1017,3 +1017,164 @@ fn glenum_to_transform_feedback_mode(value: gl::types::GLenum) -> TransformFeedb
+            ShaderStage::TessellationControl => gl::TESS_CONTROL_SHADER,
+            ShaderStage::TessellationEvaluation => gl::TESS_EVALUATION_SHADER,
+            ShaderStage::Geometry => gl::GEOMETRY_SHADER,
+            // Compute => gl::COMPUTE_SHADER,
+        }
+    }
+}
+
+fn get_supported_shader_stages(ctxt: &mut CommandContext) -> Vec<ShaderStage> {
+    let mut stages = vec![ShaderStage::Vertex, ShaderStage::Fragment];
+    if ctxt.extensions.gl_arb_tessellation_shader {
+        stages.push(ShaderStage::TessellationControl);
+        stages.push(ShaderStage::TessellationEvaluation);
+    }
+    if ctxt.extensions.gl_arb_geometry_shader4 {
+        stages.push(ShaderStage::Geometry);
+    }
+    stages
+}
+
+/// Returns the data associated with a programs subroutines.
+pub unsafe fn reflect_subroutine_data(ctxt: &mut CommandContext, program: Handle)
+                               -> SubroutineData
+{
+    if !ctxt.extensions.gl_arb_shader_subroutine {
+        return SubroutineData {
+            location_counts: HashMap::with_capacity(0),
+            subroutine_uniforms: HashMap::with_capacity(0),
+        }
+    }
+
+    let program = match program {
+        // subroutines not supported.
+        Handle::Handle(_) => return SubroutineData {
+            location_counts: HashMap::with_capacity(0),
+            subroutine_uniforms: HashMap::with_capacity(0),
+        },
+        Handle::Id(id) => id
+    };
+
+    let shader_stages = get_supported_shader_stages(ctxt);
+    let mut subroutine_uniforms = HashMap::new();
+    let mut location_counts = HashMap::new();
+    for stage in shader_stages.iter() {
+        let mut location_count: gl::types::GLint = mem::uninitialized();
+        ctxt.gl.GetProgramStageiv(program, stage.to_gl_enum(), gl::ACTIVE_SUBROUTINE_UNIFORM_LOCATIONS, &mut location_count);
+        location_counts.insert(*stage, location_count as usize);
+        let mut subroutine_count: gl::types::GLint = mem::uninitialized();
+        ctxt.gl.GetProgramStageiv(program, stage.to_gl_enum(), gl::ACTIVE_SUBROUTINE_UNIFORMS, &mut subroutine_count);
+        for i in 0..subroutine_count {
+            // Get the name of the uniform
+            let mut uniform_name_tmp: Vec<u8> = vec![0; 64];
+            let mut name_len: gl::types::GLsizei = mem::uninitialized();
+            ctxt.gl.GetActiveSubroutineUniformName(program, stage.to_gl_enum(), i as gl::types::GLuint, uniform_name_tmp.len() as gl::types::GLint,
+                                                   &mut name_len, uniform_name_tmp.as_mut_ptr() as *mut gl::types::GLchar);
+            uniform_name_tmp.set_len(name_len as usize);
+            // Get the location of the uniform
+            // TODO Not sure if this is completely correct. Do we have to append '\0' ?
+            let location = ctxt.gl.GetSubroutineUniformLocation(program, stage.to_gl_enum(), uniform_name_tmp.as_ptr() as *const gl::types::GLchar);
```

Comment:
I think this is wrong. `name_len` contains the length of the string, excluding the null terminator. So when you use `set_len()` above you're removing the null terminator from the vec.

A simple fix is to call `set_len` after calling `GetSubroutineUniformLocation`.